File: lib/src/main/java/com/sd/lib/chain/FChain.java
package com.sd.lib.chain;

import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

public class FChain {
    /** 节点链 */
    private final List<Node> mListNode = new ArrayList<>();
    private final Handler mHandler = new Handler(Looper.getMainLooper());

    /** 当前执行的节点位置 */
    private int mCurrentIndex = -1;
    /** 当前执行的节点 */
    private Node mCurrentNode = null;
    /** 是否已经通知了{@link #onFinish()} */
    private boolean mHasNotifyFinish = false;

    /**
     * 节点数量
     */
    public final synchronized int size() {
        return mListNode.size();
    }

    /**
     * 返回当前节点
     */
    @Nullable
    public final synchronized Node getCurrentNode() {
        return mCurrentNode;
    }

    /**
     * 添加节点，一个节点对象只能添加到一个链上。
     */
    public void add(@NonNull Node node) {
        synchronized (FChain.this) {
            node.init(FChain.this, mHandler);
            mListNode.add(node);
        }
    }

    /**
     * 开始
     *
     * @return true-本次调用触发了开始
     */
    public boolean start() {
        synchronized (FChain.this) {
            if (mListNode.isEmpty()) return false;
            if (mCurrentNode != null) return false;

            final int index = 0;
            final Node node = mListNode.get(index);
            if (node.getState() != NodeState.None) {
                throw new RuntimeException("Illegal node state " + node.getState());
            }

            // 重置为false
            mHasNotifyFinish = false;

            mCurrentIndex = index;
            mCurrentNode = node;
            mCurrentNode.notifyRun();
            return true;
        }
    }

    /**
     * 取消当前节点，并清空所有节点。
     */
    public void cancel() {
        synchronized (FChain.this) {
            if (mCurrentNode == null) return;

            mCurrentNode.notifyCancel();
            mCurrentNode = null;
            mCurrentIndex = -1;
            mListNode.clear();

            notifyOnFinishLocked();
        }
    }

    private synchronized void runNextNode() {
        final Node currentNode = mCurrentNode;
        if (currentNode == null) {
            // 还未开始，不允许调用此方法
            return;
        }

        if (currentNode.getState() != NodeState.Finish) {
            throw new RuntimeException("Current node has not finished " + currentNode.getState());
        }

        final int nextIndex = mCurrentIndex + 1;
        if (nextIndex >= mListNode.size()) {
            notifyOnFinishLocked();
            return;
        }

        mCurrentIndex = nextIndex;
        mCurrentNode = mListNode.get(nextIndex);
        mCurrentNode.notifyRun();
    }

    private void notifyOnFinishLocked() {
        if (mHasNotifyFinish) return;
        mHasNotifyFinish = true;

        mHandler.post(new Runnable() {
            @Override
            public void run() {
                onFinish();
            }
        });
    }

    /**
     * 结束回调，在主线程触发。
     */
    protected void onFinish() {
    }

    public enum NodeState {
        None,
        Run,
        Finish,
    }

    public static abstract class Node {
        private FChain _chain;
        private Handler _handler;
        private volatile NodeState _state = NodeState.None;
        private volatile boolean _hasRun;

        /**
         * 节点状态
         */
        @NonNull
        public final NodeState getState() {
            return _state;
        }

        private synchronized void init(@NonNull FChain chain, @NonNull Handler handler) {
            if (_chain == null) {
                _chain = chain;
                _handler = handler;
            } else {
                throw new IllegalArgumentException("Node has been added to " + _chain);
            }
        }

        private synchronized void checkInit() {
            if (_chain == null) {
                throw new RuntimeException("Node has not been added to any chain.");
            }
            if (_handler == null) {
                throw new RuntimeException("Node's handler is null.");
            }
        }

        private void notifyRun() {
            checkInit();

            synchronized (Node.this) {
                if (_state == NodeState.None) {
                    _state = NodeState.Run;
                } else {
                    throw new RuntimeException("Illegal node state " + _state);
                }
            }

            _handler.post(new Runnable() {
                @Override
                public void run() {
                    // 检查状态，如果未被取消，则通知回调
                    if (_state == NodeState.Run) {
                        _hasRun = true;
                        onRun();
                    }
                }
            });
        }

        private void notifyCancel() {
            checkInit();

            boolean stateChanged = false;
            synchronized (Node.this) {
                if (_state != NodeState.Finish) {
                    _state = NodeState.Finish;
                    stateChanged = true;
                }
            }

            if (!stateChanged) {
                // 状态未改变，不处理
                return;
            }

            if (_hasRun) {
                _handler.post(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            onCancel();
                        } finally {
                            onFinish();
                        }
                    }
                });
            }
        }

        /**
         * 执行下一个节点
         */
        protected final void nextNode() {
            checkInit();

            boolean stateChanged = false;
            synchronized (Node.this) {
                if (_state == NodeState.Run) {
                    _state = NodeState.Finish;
                    stateChanged = true;
                }
            }

            if (stateChanged) {
                if (_hasRun) {
                    _handler.post(new Runnable() {
                        @Override
                        public void run() {
                            onFinish();
                        }
                    });
                }

                // 通知下一个节点
                _chain.runNextNode();
            }
        }

        /**
         * 节点执行回调，主线程触发
         */
        protected abstract void onRun();

        /**
         * 节点取消回调，主线程触发。
         * 如果{@link #onRun()}未触发过，则此方法不会被触发。
         */
        protected void onCancel() {
        }

        /**
         * 节点结束回调，主线程触发。
         * 如果{@link #onRun()}未触发过，则此方法不会被触发。
         */
        protected void onFinish() {
        }
    }
}

File: lib/src/main/java/com/sd/lib/chain/FChain.kt
package com.sd.lib.chain

import android.os.Handler
import android.os.Looper
import java.util.concurrent.atomic.AtomicBoolean

open class FChain {
    /** 所有节点  */
    private val _nodes: MutableList<Node> = mutableListOf()

    /** 当前执行的节点位置  */
    private var _currentIndex: Int = -1
    /** 当前执行的节点  */
    private var _currentNode: Node? = null

    private val _handler = Handler(Looper.getMainLooper())

    /**
     * 是否已经启动
     */
    @Synchronized
    fun isStarted(): Boolean = _currentNode != null

    /**
     * 节点数量
     */
    @Synchronized
    fun size(): Int = _nodes.size

    /**
     * 添加节点，一个节点对象只能被添加一次
     */
    @Synchronized
    fun add(node: Node) {
        node.init(this@FChain, _handler)
        _nodes.add(node)
    }

    /**
     * 开始
     * @return true-本次调用触发了开始
     */
    @Synchronized
    fun start(): Boolean {
        if (isStarted()) return false
        if (_nodes.isEmpty()) return false

        val firstNode = _nodes[0].also {
            _currentIndex = 0
            _currentNode = it
        }

        _handler.post { onStart() }
        firstNode.notifyRun()
        return true
    }

    /**
     * 取消，并清空所有节点。
     */
    @Synchronized
    fun cancel() {
        val node = _currentNode ?: return
        node.notifyCancel()
        finish()
    }

    @Synchronized
    private fun runNextNode() {
        val node = _currentNode ?: return
        node.checkFinish()

        val nextIndex = _currentIndex + 1
        if (nextIndex >= _nodes.size) {
            finish()
            return
        }

        _nodes[nextIndex].also {
            _currentIndex = nextIndex
            _currentNode = it
        }.notifyRun()
    }

    private fun finish() {
        _currentIndex = -1
        _currentNode = null
        _nodes.clear()
        _handler.post { onFinish() }
    }

    /**
     * 开始回调，主线程触发
     */
    protected open fun onStart() = Unit

    /**
     * 结束回调，主线程触发
     */
    protected open fun onFinish() = Unit

    private enum class NodeState {
        None,
        Run,
        Finish
    }

    abstract class Node {
        private val _init = AtomicBoolean(false)

        private lateinit var _chain: FChain
        private lateinit var _handler: Handler

        private var _state = NodeState.None
        private var _hasRun = false

        internal fun init(chain: FChain, handler: Handler) {
            if (_init.compareAndSet(false, true)) {
                _chain = chain
                _handler = handler
            } else {
                error("Node has been initialized.")
            }
        }

        internal fun checkFinish() {
            check(_state == NodeState.Finish) { "Node has not finished with state:${_state}" }
        }

        internal fun notifyRun() {
            check(_state == NodeState.None) { "Require state None." }
            _state = NodeState.Run
            _handler.post {
                synchronized(_chain) {
                    (_state == NodeState.Run).also {
                        if (it) _hasRun = true
                    }
                }.let { notify ->
                    if (notify) {
                        onRun()
                    }
                }
            }
        }

        internal fun notifyCancel() {
            if (_state != NodeState.Finish) {
                _state = NodeState.Finish
                if (_hasRun) {
                    _handler.post { onCancel() }
                    _handler.post { onFinish() }
                }
            }
        }

        /**
         * 执行下一个节点
         */
        protected fun nextNode() {
            check(_init.get()) { "Node has not been initialized." }
            synchronized(_chain) {
                when (_state) {
                    NodeState.None -> error("Can not call nextNode() before onRun().")
                    NodeState.Finish -> return
                    NodeState.Run -> {
                        _state = NodeState.Finish
                        if (_hasRun) {
                            _handler.post { onFinish() }
                        }
                        _chain.runNextNode()
                    }
                }
            }
        }

        /**
         * 节点执行回调，主线程触发
         */
        protected abstract fun onRun()

        /**
         * 节点取消回调，主线程触发，
         * 如果[onRun]未触发过，则此方法不会被触发
         */
        protected open fun onCancel() = Unit

        /**
         * 节点结束回调，主线程触发，
         * 如果[onRun]未触发过，则此方法不会被触发
         */
        protected open fun onFinish() = Unit
    }
}